src/transforms/pipelines/config.rs

```diff
@@ -42,65 +52,167 @@ impl Clone for PipelineConfig {
     }
 }
 
+#[async_trait::async_trait]
+#[typetag::serde(name = "pipeline")]
+impl TransformConfig for PipelineConfig {
+    async fn build(&self, ctx: &TransformContext) -> crate::Result<Transform> {
+        let condition = if let Some(config) = &self.filter {
+            Some(config.build(&ctx.enrichment_tables)?)
+        } else {
+            None
+        };
+
+        // Setup the interior transform chain in this pipeline.
+        //
+        // NOTE: As this pipeline is not expanding to separate transforms we
+        // lose validation that happens at the topology layer (e.g. making sure
+        // input/output types of connected transforms are compatible). This is
+        // an unfortunate consequence of having a "topology in a topology".
+        //
+        // In the future, it would be great to decouple that type of validation
+        // from the physical task boundaries. You could imagine a future where
+        // we replace transform expansion with the ability to return different
+        // types of compound transforms directly from TransformConfig::build,
+        // and those know how to play nice with a subsequent graph-based
+        // validation scheme.
+        //
+        // For now, considering that most everything is log to log with regard
+        // to pipelines we avoid adding duplicate validation here in favor of
+        // future work.
+        let mut transforms = Vec::with_capacity(self.transforms.len());
+        for config in &self.transforms {
+            let transform = match config.build(ctx).await? {
+                Transform::Function(transform) => Box::new(transform),
+                Transform::Synchronous(transform) => transform,
+                _ => panic!("non-sync transform in pipeline: {:?}", config),
```

Comment:
I'm not sure if there's currently anything stopping anyone from manually writing a config with a task transform as part of a pipeline, so maybe it would be better to return a build error here.

Reply:
Oops, yes, 100%.

Reply:
Addressed in [4e1928a](https://github.com/vectordotdev/vector/pull/12602/commits/4e1928a9e4cec95d2767730e24fb58ca3be86823)